Re-enabled Weather Sensors for the ISY component.

rmkraus opened the pull request:
**Description:**
As the ISY component had been updated to support newer components and
have better UOM support, the support for ISY’s weather module was
dropped. This adds that support back into Home Assistant.

rmkraus committed:
Re-enabled Weather Sensors for the ISY component.

As the ISY component had been updated to support newer components and
have better UOM support, the support for ISY’s weather module was
dropped. This adds that support back into Home Assistant.

balloob commented on homeassistant/components/isy994.py:
```diff
@@ -132,6 +134,20 @@ def _categorize_programs() -> None:
 
+def _categorize_weather() -> None:
+    """Categorize the ISY994 weather data."""
+    global WEATHER_NODES
```
Instead of `global`, can you store it in `hass.data`. It's a dict that stores global data for the duration of hass being on.

rmkraus replied:
Oooh, that's new, and useful. For the time being, I'd like to keep it the way it is because there are a couple other globals used in this way. For the purposes of a hotfix, I'd rather not change more than I have to. Once this is in dev and master, I'd be happy to fix the ISY component to do that.

rmkraus replied:
By the by, after pushing this PR to dev, I was hoping to open a PR against the master. I have a couple of hotfixes. This one and the event decorators are broken.

balloob replied:
I was actually considering removing the event decorators because `partial` does not work with async. I also think that you're the only one using them ;-)

rmkraus replied:
Yeah, I think I am the only one too. I'll probably just convert my custom components to real components so I wont need them anymore.

Teagan42 commented on homeassistant/components/sensor/isy994.py:
```diff
@@ -309,3 +312,47 @@ def unit_of_measurement(self) -> str:
+        """Return the value of the node."""
+        # pylint: disable=protected-access
+        val = self._node.status._val
+        raw_units = self._node.uom
```
Pretty sure `uom` is a list of strings

Teagan42 replied:
No I'm definitely sure. It comes in as a list of units because `uom` could be unit of measure or enumerated device states (on, off, open, closed, etc.). If the device has one element in the `uom` list then it's a unit of measurement.

rmkraus replied:
Not this time. The weather data in the ISY library isn't stored in the same Node class. In order to make the ISYDevice abstract class to mostly work with weather data, I create a named tuple with the data. In that named tuple, UOM is just a string.

balloob commented on homeassistant/components/isy994.py:
```diff
@@ -132,6 +134,20 @@ def _categorize_programs() -> None:
+    """Categorize the ISY994 weather data."""
+    global WEATHER_NODES
+
+    items = [item for item in dir(ISY.climate)
```
It would be a lot more performant if you would cache `dir(ISY.climate)`.

balloob replied:
And while rewriting this, you could merge it with the for loop below.

balloob commented on homeassistant/components/isy994.py:
```diff
@@ -132,6 +134,20 @@ def _categorize_programs() -> None:
+    items = [item for item in dir(ISY.climate)
+             if item + '_units' in dir(ISY.climate)]
+
+    weather_node = namedtuple('WeatherNode', ('status', 'name', 'uom'))
```
Better to move this to the top

rmkraus committed:
Cleanup of the ISY Weather support.

Cleaned up the for loops used to generate nodes representing weather
data from the ISY. Moved the weather_node named tuple to the top of the
file so that it can be more easily identified.

balloob committed:
Update isy994.py

balloob commented:
I don't think that it is hotfix worthy since it must have been broken for a while? Might as well wait for 0.36 which is scheduled for a week + change from now.

rmkraus commented:
That works for me. Probably was broken for a while, so that is fair. Thanks.
